rfcs/0009_Unified_Static_Memory_Planning.md

```diff
@@ -0,0 +1,473 @@
+    #define TVM_MY_MODEL_FLASH_PARAMETER_BUFFER_SIZE xxxx
+
+    typedef struct {
+       uint8_t* dtcm;
+       uint8_t* sram;
+    }  TVMWorkspaces_my_model;
+
+    typedef struct {
+       uint8_t* itcm;
+       uint8_t* flash;
+    }  TVMParameters_my_model;
+
+    typedef struct {
+       uint8_t* input0;
+    }  TVMInputs_my_model;
+
+    typedef struct {
+       uint8_t* output0;
+    }  TVMOutputs_my_model;
+```
+### User Application
+```
+    // The User Application 
+        extern  const TVMModel my_model;
+        __attribute__((section( "ITCM" )  const uint8_t   my_model_params_1[TVM_MY_MODEL_ITCM_PARAMETER_BUFFER_SIZE] = <param_1_data>;
+        __attribute__((section( "FLASH" ), aligned( 16 )))  const uint8_t my_model_params_2[TVM_MY_MODEL_FLASH_PARAMETER_BUFFER_SIZE] = <param_2_data>;
+        __attribute__((section( "DTCM" )  static uint8_t workspace_buffer_1[TVM_MY_MODEL_DTCM_WORKSPACE_BUFFER_SIZE];
+        __attribute__((section( "SRAM" ), aligned( 16 )))  static uint8_t workspace_buffer_2[TVM_MY_MODEL_SRAM_WORKSPACE_BUFFER_SIZE];
+
+    int main(...) {
+         ...
+         TVMContext context;
+         TVMInputs_my_model_1 inputs = {input};
+         TVMOutputs_my_model_1 outputs = {output};
+         TVMWorkspaces_my_model workspaces = {
+             .sram = &workspace_buffer_1,
+             .dtcm = &workspace_buffer_2,
+         };
+         TVMParameters_my_model parameters = {
+             .flash = &my_model_params_1,
+             .itcm = &my_model_params_2
+         };
+         TVMSetWorkspaces(&context, &workspaces);
+         TVMSetParameters(&context, parameters);
+         TVMExecute(&my_model, &inputs, &outputs, &context);
+    }
+```
+# Reference-level explanation
+
+## Overview
+
+This should be a IRModule (TIR) → IRModule (TIR) pass.
+
+Inputs : 
+* AoT TIR PrimFunc ( the control function describing the call graph to operators)
+* All Operator Functions
+* the maximum size for each pool We could use "pinned_memory" (see below) to tag buffers with suggested priority order determined by the scheduler.
+
+The idea is USMP will try to pool them using the preferred "pinned_memory" and fallback whenever the size is exceeding the user provided max size for each pool (if any)
+
+Outputs : 
+* AoT TIR PrimFunc accepting pool buffers from the user.
+* All Operator functions accepting pool buffers.
+  * Each operator function should address using the correct offset in the correct pool buffer
+
+Special Parametric Inputs : 
+* function : The algorithm to be used for planning From a component PoV, the algorithm is a special input with a defined interface.
+
+The current proposal for the interface of the memory planning algorithm is as follows :
+```
+    struct BufferInfo {
+        String name_hint; // this is the tir.buffer name
+        Integer size_bytes;
+        Integer alignment;
+        Array<BufferInfo> conflicts; //the conflicting bufferinfo objs
+        Array<Integer> pool_candidates;`
```

Comment:
Does the memory planning algorithm have to assign each buffer to a pool in `pool_candidates`? Is this information not used after memory planning?

Reply:
As of now, once they are assigned with pool and offset, pool_candidates may not be used.

Correction : Array<Integer> pool_candidates --> Array<String> pool_candidates.

Reply:
it does seem like it would be useful to split the pre-memory-planning data structure apart from the post-memory planning data structure. perhaps:
- an attribute `"memory_constraints"` is expected on nodes as input to memory planning
- an attribute `"memory_placement"` is expected on nodes as an output

one concern with this approach is it may be slower if we need to copy all nodes involved. cc @jroesch for opinions on how to handle this sort of thing in the new TE-compiler world.